### DISCUSSION TITLE
Veien videre for Lerna og Yarn

### DISCUSSION THREAD

**Discussion by wkillerud** [fremtind/jokul#2626] [Annet]
## Tingenes tilstand 

Som dere kanskje har fått med dere er hverken [Yarn 1.x](https://github.com/yarnpkg/yarn/blob/master/CHANGELOG.md#12211) eller [Lerna](https://github.com/lerna/lerna/issues/2703#issuecomment-742107407) under aktiv utvikling (edit: Lerna har fått en ny maintainer!). Vi er avhengig i dag av spesielt Lerna for å gjøre versjonering og publisering av pakker, og oppdatering av Changelogs, på en automatisert måte. Lerna fungerer ikke med nyere versjoner av Yarn.

Siden Yarn 1.x har det skjedd ting både i Yarn, med npm, og med alternativer som pnpm. Alle tre har tilsynelatende støtte for "workspaces" i en eller annen form. De kan med andre ord ta over noen av oppgavene vi bruker Lerna til i dag. En migrering vekk fra dagens situasjon trenger altså ikke være 1:1 med tanke på hvilke verktøy som gjør hvilke oppgaver.

For meg er det ikke noe alternativ på sikt å bli værende på de versjonene vi har i dag.

## Oppgaver som må løses

<table>
<thead>
<tr>
	<th>Oppgave</th>
	<th>Løst i dag av</th>
</tr>
</thead>
<tbody>
<tr>
	<td>Installering og oppdatering av avhengigheter (package manager)</td>
	<td>Yarn Classic 1.x</td>
</tr>
<tr>
	<td>Lenking mellom pakker i monorepoet</td>
	<td>Lerna</td>
</tr>
<tr>
	<td>Kjøre kommandoer på tvers av pakker i monorepoet (task runner)</td>
	<td>Lerna</td>
</tr>
<tr>
	<td>Automatisk versjonere pakker basert på endringer som er gjort siden sist</td>
	<td>Lerna, Commitizen, Commitlint</td>
</tr>
<tr>
	<td>Automatisk oppdatere changelogs med endringer</td>
	<td>Lerna, Commitizen, Commitizen Lerna Changelog, Lerna Clean Changelog CLI, Commitlint</td>
</tr>
<tr>
	<td>Automatisk publisere endringer</td>
	<td>Lerna, Commitizen, Commitizen Lerna Changelog, Lerna Clean Changelog CLI, Commitlint</td>
</tr>
</tbody>
</table>

## Alternativ

Flere som har gått vekk fra Lerna har tatt i bruk https://github.com/changesets/changesets for versjonering, changelogs, og publisering.

Som en task runner har Microsoft sin [Lage](https://github.com/microsoft/lage) fått litt hype (ja, norsk for `make`).

Vi har flere alternative package managers som kan funke, hver med sine "killer features". 

## Kor ska vi reis?

- Fortsette på Yarn? Gå til npm? Droppe begge for pnpm?
- Kan Changesets fungere som en erstatter til dagens conventional commits + Lerna?
- `lage build`?

**Comment by piofinn** [comment]
> Som en task runner har Microsoft sin Lage fått litt hype (ja, norsk for make)

Typisk amerikanere å bruke infinitivsformen når de skulle brukt imperativ! "Lag" var kanskje litt belasta 😆

### Turborepo

Som jeg nevnte kjapt på utvidet standup har jeg lekt meg så vidt med [Turborepo](https://turborepo.org/). Det er en task runner for monorepo som håndterer topologisk kjøring av bygg (sørger for først å bygge pakker som andre pakker er avhengige av). Det ser ut til å funke veldig greit, og oppsettet er veldig enkelt!

Killer feature her er vel egentlig caching og diff. Har du bygget før, og ikke gjort endringer i en pakke, blir den ikke bygget på nytt. Man har også myligheten til en remote cache som gjør at dette også funker på tvers av branches/commits, slik at man bare trenger å bygge de pakkene som faktisk er endret når man kjører CI på en pull request. Det bør spare en del tid!

Jeg støtte på en snag der bygget hang seg opp på `text-input-react`-pakken, men jeg tror det kan skyldes noen dårlig definerte interne avhengigheter i repoet. Skal prøve å finne ut av det!

### standard-version

Når det kommer til versjonering har jeg nettopp jobbet litt med det i et par av de interne repoene våre, og der har jeg brukt [standard-version](https://github.com/conventional-changelog/standard-version#standard-version). Det er verktøyet som blir brukt bak scenen av lerna for å generere changelogs og sette versjonsnummer ut fra typen commits. Det funka helt smertefritt i et enkelt repo, men jeg har ikke sjekket om det er noen quirks i forhold til å bruke det direkte på et monorepo.

### Yarn

Siden Yarn ble nevnt over vil jeg bare skyte inn at selv Yarn 1.x har støtte for workspaces; det er denne funksjonaliteten som ligger i bunn av lerna for linking av pakker 😊 Det er imidlertid veldig mange andre gode grunner til å se på oppgradering til 2.0 (eller migrering til (p)npm), så det bør absolutt undersøkes!

[Plug'n'play](https://yarnpkg.com/features/pnp)-funksjonaliteten i Yarn 2 kan nok være med på å kutte ned tiden på CI/CD-flytene våre betraktelig! Og de strengere kravene til avhengigheter som følger med kan nok være med på å finne en del latente bugs i oppsettet vårt.

**Reply by wkillerud** [reply]
> Killer feature her er vel egentlig caching og diff. Har du bygget før, og ikke gjort endringer i en pakke, blir den ikke bygget på nytt. Man har også myligheten til en remote cache som gjør at dette også funker på tvers av branches/commits, slik at man bare trenger å bygge de pakkene som faktisk er endret når man kjører CI på en pull request. Det bør spare en del tid!

Nice! _Det_ hadde vært digg.

Om `standard-version`, dersom det kan funke i et monorepo enten på egenhånd eller via Turborepo så hadde det jo vært digg å slippe å rigge om hele workflowen som er etablert rundt conventional commits.

Og yeah, den "zero install"-verden Yarn 2 og 3 tegner virker ganske forlokkende. Verdt å prøve, om ikke annet.

**Reply by Mikaila94** [reply]
Jeg har ikke lest meg noe særlig på dette selv, men forslagene her virker lovende.
Jeg blir gjerne med for å lære mer om dette når dette oppgraderingsprosjektet settes i gang.

**Comment by wkillerud** [comment]
[Interessant diskusjon på Hacker News om pnpm](https://news.ycombinator.com/item?id=30919487&utm_source=pocket_mylist). Konsensus virker være at det er _pretty good_. Mange sammenligner ytelsen til pnpm med Yarn Classic og ser store forbedringer. Så ingen "rettferdige" sammenligninger med Yarn 3 vs pnpm.

Denne featuren virker kul!

> If you have a pnpm monorepo, you need to know about https://pnpm.io/filtering#--filter-since which allows to run your test/lint/etc on only the packages that have been impacted by changes from main.

**Comment by wkillerud** [comment]
En gledelig utvikling i Lerna 😄 Kan hende et nytt roadmap aligner med behovet vårt. Watch this space.
https://github.com/lerna/lerna/issues/3121
